[MapD] Add geo spatial datatype support

xmnlab opened the pull request:
Resolves partial #1665

- added support to geo spatial data

MapD GEO spatial functions support should be addressed in another PR

cpcloud commented on ibis/expr/datatypes.py:
```diff
@@ -4,6 +4,7 @@
 import numbers
 import re
 
+from shapely import geometry
```
The geospatial types should be moved into their own file, `geospatial.py` so that users who don't have `shapely` installed can still use the library.

cpcloud commented on ibis/expr/datatypes.py:
```diff
@@ -605,6 +606,140 @@ def _equal_part(self, other, cache=None):
+        self.x = x
+        self.y = y
+
+    def __str__(self):
```
Is this not sufficiently defined by the `DataType` implementation?

xmnlab replied:
I think so .. thanks!

cpcloud commented on ibis/expr/datatypes.py:
```diff
@@ -605,6 +606,140 @@ def _equal_part(self, other, cache=None):
+
+    __slots__ = 'x', 'y'
+
+    def __init__(self, x, y, nullable=True):
```
I think you're misunderstanding how `DataType`s work. `DataType`s are just type information, and shouldn't be used to construct values. The way a user would make a point object might be with a 2-tuple:

```python
ibis.literal((1, 2), type='point')
```

xmnlab replied:
oh I see .. thanks for the explanation!

cpcloud commented on ibis/expr/datatypes.py:
```diff
@@ -605,6 +606,140 @@ def _equal_part(self, other, cache=None):
+        return self.x == other.x and self.y == other.y
+
+
+assert hasattr(Point, '__hash__')
```
Remove this.

xmnlab replied:
ok! thanks!

cpcloud commented on ibis/expr/datatypes.py:
```diff
@@ -605,6 +606,140 @@ def _equal_part(self, other, cache=None):
+        return self.points == other.points
+
+
+assert hasattr(Line, '__hash__')
```
Remove this.

xmnlab replied:
ok! thanks!

cpcloud commented on ibis/expr/datatypes.py:
```diff
@@ -605,6 +606,140 @@ def _equal_part(self, other, cache=None):
+        return self.lines == other.lines
+
+
+assert hasattr(GeoSpatial, '__hash__')
```
Remove this.

xmnlab replied:
ok! thanks!

cpcloud commented on ibis/expr/datatypes.py:
```diff
@@ -929,6 +1093,21 @@ def type(self):
 
         field : [a-zA-Z_][a-zA-Z_0-9]*
+
+        # geospatial
```
Since you added a rule to the grammar, please add a rule like:

```antlr
geospatial: point | line | polygon | multipolygon
```

xmnlab commented:
@cpcloud @kszucs 

I am moving the geospatial types to it own file. now I have a question.
Inside geospatial.py I will import datatypes to create geo spatial data types. so where I should import/add the spatial data? My first guess is to add this inside datatype.py but it will create a circular import ... so I think it is not desired.

what would be the workflow for this implementation?

cpcloud commented on ibis/expr/geospatial.py:
```diff
@@ -0,0 +1,181 @@
+            else:
+                x = 0
+                y = 0
+            return Point(x, y)
```
You need to remove the arguments to the constructors here and elsewhere in the parser.

xmnlab replied:
ok! thanks!

xmnlab commented:
@cpcloud @kszucs

I moved the geospatial types to it own file. now I have a question.
Inside geospatial.py I will import datatypes to create geo spatial data types. so where I should import/add the spatial data? My first guess is to add this inside datatype.py but it will create a circular import ... so I think it is not desired.

what would be the workflow for this implementation?

kszucs commented on ibis/expr/datatypes.py:
```diff
@@ -1201,6 +1280,27 @@ def infer_null(value):
 
+# geo spatial data type
+@infer.register(tuple)
+def infer_point(value):
```
We have not overloaded infer with tuple argument yet, so theoretically We could use this for Point datatype, but We might have a tuple type in the future. IMO remove this.

kszucs commented:
@xmnlab Try removing the infer functions above.

xmnlab commented:
@kszucs thanks for the feedback!

with out infer function, it raises an error:

```
---------------------------------------------------------------------------
InputTypeError                            Traceback (most recent call last)
<ipython-input-3-26e2613b0aa2> in <module>
      1 point = (0, 0)
----> 2 l_point = ibis.literal(point, type='point')
      3 
      4 print(l_point.compile())
      5 print(type(l_point))

/mnt/sda1/dev/quansight/ibis/ibis/expr/types.py in literal(value, type)
    894         dtype = dt.null
    895     else:
--> 896         dtype = dt.infer(value)
    897 
    898     if type is not None:

/mnt/sda1/storage/miniconda/envs/ibis/lib/python3.6/site-packages/multipledispatch/dispatcher.py in __call__(self, *args, **kwargs)
    276             self._cache[types] = func
    277         try:
--> 278             return func(*args, **kwargs)
    279 
    280         except MDNotImplementedError:

/mnt/sda1/dev/quansight/ibis/ibis/expr/datatypes.py in infer_dtype_default(value)
   1189 @infer.register(object)
   1190 def infer_dtype_default(value):
-> 1191     raise com.InputTypeError(value)
   1192 
   1193 
```

xmnlab commented:
@kszucs 

I could resolve the problem here:

- added a infer function for tuple -> Array(Primitive())
- added cast for array to point, line, polygon and multipolygon

let me know if this is reasonable or if you prefer another approach.

kszucs commented:
That sounds good!

xmnlab commented:
@kszucs it seems all tests passed except for azure (mysql installation issue) and python27 (sqlite issues)

do you have any idea how to fixed that?

kszucs commented:
MySQL testing has been disabled on the master. I'm rebasing it.

xmnlab commented:
Thanks @kszucs !! It seems awesome!

kszucs commented:
@xmnlab @cpcloud In overall the new datatypes are working, however We'll need a couple of follow-up PRs. 

The current implementation doesn't reflect the hierarchy between the spatial types:
```
Point = Array[Numeric, 2]
Line = Array[Point]
Polygon = Array[Line]
Multypolygon = Array[Polygon]
```
And the implicit casting rules need to be more restrictive as well.

cpcloud commented on ibis/expr/datatypes.py:
```diff
@@ -602,6 +603,38 @@ def __str__(self) -> str:
+    __slots__ = ()
+
+
+class MultiPolygon(GeoSpatial):
```
Can you document these types? I think people may wonder how these types map to PostgreSQL's versions of these, since they are pretty similar. Also, these types should be flexible enough to support Postgres's versions. I think the exercise of going the comparison will be very helpful in determining if this is the right set of types to support databases that support them.

xmnlab replied:
it sounds good .. I will also change Line to Linestring. I will also add srid type information.

cpcloud commented on ibis/expr/tests/test_value_exprs.py:
```diff
@@ -50,26 +50,24 @@ def test_unicode():
-    ]
-)
-def test_literal_cases(value, expected_type):
+@pytest.mark.parametrize(['value', 'expected_type'], [
```
Can you leave this formatting as is?

xmnlab replied:
OK!

xmnlab replied:
for some unknown reason I lost this ... I am doing that.

cpcloud commented on ibis/expr/tests/test_value_exprs.py:
```diff
@@ -79,25 +77,41 @@ def test_literal_cases(value, expected_type):
+multipolygon1 = [polygon1, polygon2]
+
+
+@pytest.mark.parametrize(['value', 'expected_type'], [
```
Please leave this formatting.

xmnlab replied:
OK .. I am doing that. thanks!

cpcloud commented on ibis/expr/types.py:
```diff
@@ -857,18 +886,35 @@ def literal(value, type=None):
+        has_inferred = True
 
-    if type is not None:
+    if type is None:
```
Please add some comments here. It's not clear why all of this additional logic was introduced.

xmnlab replied:
ok!

cpcloud commented on ibis/mapd/operations.py:
```diff
@@ -330,9 +330,43 @@ def _cross_join(translator, expr):
 
 
+def _format_point_value(value):
+    return ' '.join([str(v) for v in value])
```
You don't need to construct a `list` when calling `str.join`.

xmnlab replied:
thanks!

kszucs replied:
@xmnlab please remove the list comprehension and use a generator instead like Phillip mentioned.

xmnlab replied:
sounds good. sorry I lost that for a unknown reason. I am fixing that again. thanks again!

cpcloud commented on ibis/expr/datatypes.py:
```diff
@@ -602,6 +603,38 @@ def __str__(self) -> str:
         )
 
 
+class GeoSpatial(DataType):
```
Is there some shared functionality among all child classes of this class? If not, then we should just make the geospatial types subclass from `DataType` and not tie them together unnecessarily.

xmnlab replied:
I thought to create the GeoSpatial data type just to generalize the functions parameters, for example: 

```
ST_Distance(poly1, ST_GeomFromText('POINT(0 0)'))
```

where st_distance returns shortest planar distance between geometries

maybe would be better to change GeoSpatial name to Geometry

what do you think?

xmnlab replied:
actually I will think more about this .. postgis use geometry and geography types ...

cpcloud commented on ibis/expr/datatypes.py:
```diff
@@ -1345,6 +1422,15 @@ def can_cast_variadic(
 
 
+# geo spatial data type
+@castable.register(Array, Point)
```
This implies that _any_ array can be cast to a point. Is that actually true?

cpcloud replied:
I see @kszucs follow-up points above, disregard.

cpcloud commented on ibis/mapd/operations.py:
```diff
@@ -330,9 +330,43 @@ def _cross_join(translator, expr):
+    elif isinstance(expr, ir.PolygonScalar):
+        return "POLYGON({0!s})".format(_format_polygon_value(value))
+    elif isinstance(expr, ir.MultiPolygonScalar):
+        return "MULTIPOLYGON({0!s})".format(_format_multipolygon_value(value))
```
You don't need the `0!s` in the format spec, just write it as `{}`.

xmnlab commented:
@cpcloud @kszucs 

I was thinking, I will change the geospatial literals to accept 2 new argument. example:

```
literal((1, 2), type='point', geotype='geographic', srid='4326')
```

when geotype and srid are optional.

xmnlab commented:
hey @kszucs @cpcloud 

any thought about the last comment (https://github.com/ibis-project/ibis/pull/1666#issuecomment-443954609) ?

xmnlab commented:
@cpcloud @kszucs any feedback?

xmnlab commented:
that is awesome! thanks @kszucs !

xmnlab commented:
@cpcloud any thought about this PR?

xmnlab commented:
@cpcloud any thought about this PR?

kszucs commented:
@xmnlab please rebase

xmnlab commented:
@cpcloud @kszucs any feedback about this PR?

kszucs commented on ibis/expr/datatypes.py:
```diff
@@ -946,6 +1043,28 @@ def type(self) -> DataType:
         struct : "struct" "<" field ":" type ("," field ":" type)* ">"
 
         field : [a-zA-Z_][a-zA-Z_0-9]*
+
```
@xmnlab please add test cases for the newly introduces parsing parts

kszucs replied:
So for example add extend the parameters of [test_dtype](https://github.com/ibis-project/ibis/blob/master/ibis/expr/tests/test_datatypes.py#L27) with `linestring;<srid>:<geotype>`

kszucs replied:
@xmnlab look the uncovered lines: https://codecov.io/gh/ibis-project/ibis/pull/1666/src/ibis/expr/datatypes.py#L1191

kszucs replied:
Codecov doesn't redirect properly, see the lines from 1191.

xmnlab replied:
Tests added

xmnlab replied:
thanks @kszucs! let me know if it is missing anything else :+1:

kszucs commented on ibis/expr/datatypes.py:
```diff
@@ -602,6 +603,67 @@ def __str__(self) -> str:
+    scalar = ir.PointScalar
+    column = ir.PointColumn
+
+    __slots__ = 'geotype', 'srid'
```
@xmnlab These slots are already defined in `GeoSpatial`, but to prevent initializing `__dict__` You still need to define `__slot__ = ()` for each descendant classes like LineString below.

xmnlab replied:
OK sounds good I am doing that now .. thanks :+1:

kszucs commented on ibis/mapd/operations.py:
```diff
@@ -331,9 +333,70 @@ def _cross_join(translator, expr):
+
+
+def _format_linestring_value(value):
+    return ', '.join([
```
Same as above, remove change list comprehension.

xmnlab replied:
ok. thanks!

kszucs commented on ibis/mapd/operations.py:
```diff
@@ -331,9 +333,70 @@ def _cross_join(translator, expr):
+    # geo spatial data type
+    if isinstance(expr, ir.PointScalar):
+        result = "POINT({0})".format(_format_point_value(value))
+        result = _format_geo_metadata(op, result)
```
Return directly.

xmnlab replied:
I am doing that. thanks!

kszucs commented:
@xmnlab please add an unreleased version to https://github.com/ibis-project/ibis/blob/a16772aea43a936cadc39046ab96d3f54526ecdf/docs/source/release.rst

With an "Experimental GeoSpatial datatype support" (or something like that) entry and a reference to the appropiate issue.

kszucs commented on ibis/__init__.py:
```diff
@@ -55,8 +53,7 @@
 
 with suppress(ImportError):
     # pip install ibis-framework[mapd]
-    if sys.version_info.major >= 3:
```
Nice catch!

kszucs commented:
The coverage is good now too. Thanks @xmnlab!
